3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/PiiPseudonymizerEnrichment.scala

```diff
@@ -0,0 +1,237 @@
+/*
+ * Copyright (c) 2017-2018 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+
+package com.snowplowanalytics
+package snowplow.enrich
+package common.enrichments.registry
+
+// Scala
+import scala.collection.JavaConverters._
+
+// Snowplow
+import com.fasterxml.jackson.databind.node.ArrayNode
+import com.jayway.jsonpath.MapFunction
+import com.snowplowanalytics.iglu.client.{SchemaCriterion, SchemaKey}
+import com.snowplowanalytics.snowplow.enrich.common.ValidatedNelMessage
+import com.snowplowanalytics.snowplow.enrich.common.utils.MapTransformer.TransformMap
+import com.snowplowanalytics.snowplow.enrich.common.utils.ScalazJson4sUtils
+
+// Iglu
+import iglu.client.validation.ProcessingMessageMethods._
+
+// Scala libraries
+import org.json4s.JValue
+import org.json4s.JsonAST._
+import org.json4s.jackson.JsonMethods
+import org.json4s.jackson.JsonMethods.{compact, parse, render}
+import org.json4s.DefaultFormats
+import org.json4s.MappingException
+
+// Java
+import java.security.{MessageDigest, NoSuchAlgorithmException}
+import com.fasterxml.jackson.databind.JsonNode
+import com.fasterxml.jackson.databind.node.{ObjectNode, TextNode}
+import com.jayway.jsonpath.spi.json.JacksonJsonNodeJsonProvider
+import com.jayway.jsonpath.{Configuration, JsonPath => JJsonPath, Option => JOption}
+
+// Scalaz
+import scalaz._
+import Scalaz._
+
+/**
+ * PiiField trait. This corresponds to a configuration top-level field (i.e. either a POJO or a JSON field) along with
+ * a function to apply that strategy to a TransformMap
+ *
+ * Members are:
+ * strategy: the strategy to be used in this field
+ * fieldName: the fieldName where this startegy will be applied
+ * transformer: a function that modifies the TransformMap by adding PII transformations where appropriate
+ */
+sealed trait PiiField {
+  def strategy: PiiStrategy
+  def fieldName: String
+  protected def applyStrategy(fieldValue: String): String
+  def transformer(transformMap: TransformMap) =
+    transformMap.collect {
+      case (inputField: String, (tf: Function2[String, String, Validation[String, String]], outputField: String))
+          if (outputField == fieldName) =>
```

Comment:
Type-erasure one is super-important. But readability one is also untouched - it takes several minutes to understand its simple purpose.